### ISSUE TITLE
iomgr objects leaking upon shutdown on Windows

### ISSUE THREAD

**Issue by jtattermusch** [grpc/grpc#8697]
There seems to be 2 endpoints leaking (client and server) for every insecure connection  and some more objects for secure connections.

```
D1109 19:05:02.746444 0 c:\users\jtattermusch\github\grpc\src\core\lib\iomgr\iomgr.c:101: Waiting for 421 iomgr objects to be destroyed
D1109 19:05:03.751882 0 c:\users\jtattermusch\github\grpc\src\core\lib\iomgr\iomgr.c:128: Failed to free 421 iomgr objects before shutdown deadline: memory leaks are likely
D1109 19:05:03.751882 0 c:\users\jtattermusch\github\grpc\src\core\lib\iomgr\iomgr.c:82: LEAKED OBJECT: client:socket=0x00AD55C8 00AD5648
D1109 19:05:03.751882 0 c:\users\jtattermusch\github\grpc\src\core\lib\iomgr\iomgr.c:82: LEAKED OBJECT: tcp_server:ipv6:[::1]:56553:socket=0x00AD5738 00AD57B8
```

**Comment by y-zeng** [comment]
There are also some iomgr object leaks on Linux.

**Comment by hsaliak** [comment]
What is the application code that triggered this? cc @dgquintas

**Comment by jtattermusch** [comment]
I believe it can reproduced by almost any Windows code - I was running a modified version of helloworld example.
The problem exhibits as shutdown taking ~5seconds (C core waiting for destroying iomgr objects, which never happens) and printing the error messages mentioned if logging is enabled.

The reason why we haven't discovered that is that besides printing the logs, and shutdown taking longer, the application finishes correctly.

**Comment by jtattermusch** [comment]
I believe this could be related to #8553. 
When debugging, tcp->refcount starts with value "2". Upon shutdown, it never drops below 1.

**Comment by dgquintas** [comment]
@sreecha has been doing some work lately that may be related for a fix

**Comment by sreecha** [comment]
fyi: What I am working on would not fix this issue..

**Comment by ctiller** [comment]
Oh, that ref should probably be initialized to 1.

**Comment by jtattermusch** [comment]
@ctiller  yeah, I figured that out yesterday and it seems to work well. Fix is in #8827.
